### DISCUSSION TITLE
inaccurate DWI masks and estimated distortion maps after denoising data with patch2self

### DISCUSSION THREAD

**Discussion by psadil** [dipy/dipy#2583] [Preprocessing]
Hello!
     I'm trying to track down the source of a failure in preprocessing that seems related to `patch2self`, but I'm not sure whether it's `patch2self` specifically or some deeper issue with data quality. 

I'm using `dipy`'s `patch2self`through `QSIprep` to help process about 400 multishell scans. The scans include b0's with opposite phase encoding, and `topup` is enabled. When using `patch2self`, the DWI brain mask fails in about half of the participants (that is, the mask ends up as a rectangular region of the DWI that covers the entire image). In most of these scans, it looks like `topup` is also failing to estimate distortion well. For example, here is the distortion map plotted on top of one of the uncorrected b0 images. 

![image](https://user-images.githubusercontent.com/8172767/165313185-e0d73667-ed8a-474e-b164-869f9175a4aa.png)

However, using other denoising options (`dwidenoise` from MRtrix, or not denoising at all) always produces more typical distortion maps and DWI brain masks (at least, given a test on 20 participants). So, it seems like `patch2self` contributes to the processing failures, but again it's not clear whether `patch2self` is just revealing issues in quality. 

I haven't noticed obvious differences between scans that end up with failed masks and scans with good masks, based on visual inspection and plots of the QC metrics produced by `QSIprep` (e.g., number bad slices, mean fd). But I also don't have much experience with DWI, so I could be missing something obvious. 

I don't necessarily need to use `patch2self`, but I would like to understand how it might be contributing to the failed masks and distortion estimation. Do you have recommendations for troubleshooting? 

Let me know if there would be a better place to send this question. 

Thanks for the help!

```
singularity run -e --nv docker://pennbbl/qsiprep:0.14.2 \ # <--- also with 0.14.3
  "${BIDS}" . participant \
  --output-resolution 1.7 \
  --denoise-method patch2self --unringing-method mrdegibbs \
  --hmc_model eddy --eddy_config eddy_params.json \
  --fs-license-file license.txt -w $PWD --participant_label "${subject_id}" -vvv​
```

**Comment by ShreyasFadnavis** [comment]
Hi @psadil ! Thanks for reaching out. I think Patch2Self is definitely flagging subjects where certain volumes are not good, due to which the masking is failing (I have seen this issue on some other data in the past). 

Since you are using QSIPrep, maybe you can use the T1w masks? They should be there in the `/anat` folder of the QSIPrep outputs. QSIPrep registers everything in the same space. 

Another workaround for this issue is that you can do the masking before Patch2Self is done on the data. I will have to defer to @mattcieslak to know how easy it is to do this? 

Let me know if you need help with either of these!

Thanks,
Shreyas

**Reply by psadil** [reply]
Thanks for the reply! Yes, we might be able to use the T1w masks, or do something like applying `patch2self` only after QSIprep is run with no denoising. 

> I think Patch2Self is definitely flagging subjects where certain volumes are not good, due to which the masking is failing (I have seen this issue on some other data in the past).

Ah, unfortunate. The data describe a population that has been tricky to scan, so issues in quality wouldn't be too surprising.  Though, as I said, we haven't seen obvious differences in the distribution of, e.g., number of bad slices in those participants with failing vs successful masks. Do you have thoughts on a metric we might be able to calculate that would be predictive of poor masking, or which could identify the problematic volumes?

**Reply by ShreyasFadnavis** [reply]
I think the issue is limited to masking only. Patch2Self should be able to handle bad slices, random signal dropouts and outliers in the data. 

I like @mattcieslak's suggestion -- You can just run patch2self on each subject after QSIPrep is finished. In DIPY, we have a command line too for Patch2Self which can make this really simple to do: `dipy_denoise_patch2self your_dwi.nii.gz you_bvals.bval`. More options can be found by running `dipy_denoise_patch2self -h` 👍🏽

**Comment by mattcieslak** [comment]
We're going to probably change the way QSIPrep handles patch2self and have it run after motion/distortion/eddy current correction. If you'd like to use p2s for now, I'd recommend using none as the denoising method and using patch2self on the desc-preproc_dwi file

**Reply by psadil** [reply]
@mattcieslak,, that's good to know. Thanks.  Let me know if there's something I can test; I'd be happy to try out different `QSIprep` configurations.
